### DISCUSSION TITLE
How to obtain Ref when in global function as from isolate or background process (eg: workmanager)

### DISCUSSION THREAD

**Discussion by peterweb2005** [rrousselGit/riverpod#1956] [General]
as title

**Comment by TimWhiting** [comment]
The state in an isolate is by definition **Isolated**. So you won't have access to the same state in your flutter app. However you can use a raw `ProviderContainer` along with providers to manage the state in your isolate. The providers then would have access to a `Ref`. Riverpod supports this pure dart state management quite well.

**Comment by peterweb2005** [comment]
Thanks,
really want to use just as container when in Bg global,
actually seen the discussion flag some like without widget, and ProviderContainer

Thanks

**Comment by TimWhiting** [comment]
There are a few usage patterns where you actually could share the same container you use for pure dart code with the widget tree. For functions that **have** to be global for example you could use the following:

```dart
void someFunction(ProviderContainer c){
  c.read(someProvider);
}

void main() {
  final c = ProviderContainer();
  someFunction(c);
  runApp(UncontrolledProviderScope(container: c, child: const MyApp(),));
  c.dispose();
}
```

However, it is best to structure your app so that this is not necessary. Most dart code that wants to have access to a `Ref` should be in a provider of some kind.

This approach does not work with isolates though as already mentioned.

If this answers your question feel free to close the issue, if you still have questions maybe give a simple example of what you want to do.

**Comment by peterweb2005** [comment]
after i used the provider container,
experienced many 
'listenable._origin != origin': a provider cannot depend on itself

edit
just 1 error
structure: just 1 global simple provider, store a simple class, holding global values,
then on error, read the context would show the error
